doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+
+```c
+ElektraKey * k1 = elektraKeyNew (&(ElektraKeyname){
+    .ns = ELEKTRA_NS_SYSTEM,
+    .name = "foo\0bar\0baz",
+    .size = 14
+});
+ElektraKeyset * ks1 = elektraKeysetNew (8);
+elektraKeyInsertAndRelease (ks1, k1);
+
+ElektraKey * i1 = elektraKeysetGet (ks1, 0);
+
+// WARNING USAFE OPERATIONS
+elektraKeyUnlockName (i1); // name can now be modified even though k1 is still part of ks1
+
+// WARNING ILLEGAL OPERATIONS
+
+// because of the elektraKeyUnlockName, this is now an ILLEGAL operation, we are changing the name of a key that is part of a keyset
+// without elektraKeyUnlockName it would be safe, and elektraKeySetName would return an error
+elektraKeySetName (i1, name2);
+
+
+// having too many elektraKeyRelease is also an IILEGAL operation
+// here we've already given up our references to these keys, so calling elektraKeyRelease again means we're releasing somebody else's reference
+// in this case this will even lead to the memory of k1 being freed and ks1 containing invalid pointers, because we're releasing the reference to k1 that is held by ks1
+elektraKeyRelease (k1);
+```
+
+## API outside `libelektra-core`
+
+### Names
+
+```c
+void elektraKeynameAppend (ElektraKeyname * name, const char * suffix, size_t size);
+
+// constness of return value follows constness of name
+ELEKTRA_CONST char * elektraKeynameLastPart (const ElektraKeyname * name, size_t * outSize);
+void elektraKeynameReplaceLastPart (ElektraKeyname * name, const char * newPart);
+
+// constness of return value follows constness of name
+ELEKTRA_CONST char * elektraKeynameNextPart (const ElektraKeyname * name, const char * currentPart);
+
+/**
+ * allocates a new string with the escaped form of @p name
+ */
+char * elektraKeynameEscape (const ElektraKeyname * name);
+
+/**
+ * Verifies that @p escaped is a valid escaped name, and optionally produces
+ *   (1) the canonical escaped form
+ *   (2) the size of the canonical escaped form
+ *   (3) the unescaped form
+ *   (4) a pointer to the first error in @p escaped
+ *
+ * @param dest If != NULL, will contain newly allocated unescaped form on success
+ * @param escaped Escaped name to process (NUL-terminated string)
+ * @param canonial If != NULL, will be newly allocated string (NUL-terminated) with canonical form of @p escaped
+ * @param canonicalSize If != NULL, will be the length of the string written to @p canonical (including NUL-terminator)
+ * @param errorPtr If != NULL, will be a pointer into @p escaped, which points to the location of the first error, i.e. the error which is described by the return value
+ *
+ * @returns 0 if @p escaped is a valid escaped name (other values for different kinds of errors)
+ */
+ElektraErrorCode elektraKeynameProcessEscaped (
+    ElektraKeyname * dest,
+    const char * escaped,
+    char ** canonical,
+    size_t * canonicalSize,
+    const char ** errorPtr
```

Comment:
I hope we can do better. (Unless too many functions are too much size-overhead.)

Reply:
How do you want to do better? What is your problem with this function? Do you have any concrete ideas?

Just saying "I hope we can do better" helps nobody...

Reply:
We first need to benchmark the too-many-functions "problem". First the problem must be clear, then we can talk about solutions.

Reply:
The benchmark would be to compare:

- a version with very few functions and many arguments each, e.g. `elektraKeySetBit(int operation, int whatDoLock, ...)`
- a version with a minimal set of functions which have only little arguments each (more or less what we have now in the case of locking the key, asking for the sync bit etc.)

At the moment we are a bit inconsistent about this more functions vs. more arguments.

Reply:
No general conclusion needed. It is simply about what is better for Elektra, with the constraint that the core stays small in size.